### PR TITLE
codeintel: fix nil pointer in retention policy match overview for visible to tip-of-default-branch uploads

### PR DESCRIPTION
**Pull request by Strum355**
Addresses bug when a non tip-of-default-branch upload was retained by being visible to a tip-of-default-branch commit, causing nil pointer as an incorrect assumption was made (see deleted comment)
 
Also updated UI to limit number of visible commits listed.

## Test plan

Added unit tests and manually tested in UI

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -16,6 +16,7 @@ All notable changes to Sourcegraph are documented in this file.
 ### Added
 
 - Pings now include code host integration usage metrics [#31379](https://github.com/sourcegraph/sourcegraph/pull/31379)
+- LSIF upload pages now include a section listing the reasons and retention policies resulting in an upload being retained and not expired. [#30864](https://github.com/sourcegraph/sourcegraph/pull/30864)
```
Comment (efritz) [inline CHANGELOG.md:19]: :nice:

File: client/web/src/enterprise/codeintel/uploads/components/UploadRetentionStatusNode.tsx [hunk 2]
```diff
@@ -48,8 +48,12 @@ const RetentionPolicyRetentionMatchNode: FunctionComponent<{ match: RetentionPol
                     Retained: {match.matches ? 'yes' : 'no'}
                     {match.protectingCommits.length !== 0 && (
                         <>
-                            , by visible {pluralize('commit', match.protectingCommits.length)}{' '}
-                            {match.protectingCommits.map(hash => hash.slice(0, 9)).join(', ')}
+                            , by {match.protectingCommits.length} visible{' '}
+                            {pluralize('commit', match.protectingCommits.length)}, including{' '}
+                            {match.protectingCommits
+                                .slice(0, 4)
+                                .map(hash => hash.slice(0, 9))
```
Comment (efritz) [inline client/web/src/enterprise/codeintel/uploads/components/UploadRetentionStatusNode.tsx:55]: Where do 4 and 9 come form here?

File: enterprise/cmd/frontend/internal/codeintel/resolvers/upload_retention.go [hunk 3]
```diff
@@ -129,19 +129,22 @@ func (r *resolver) populateMatchingCommits(
 		if policyMatches, ok := matchingPolicies[commit]; ok {
 			for _, policyMatch := range policyMatches {
 				if policyMatch.PolicyDuration == nil || now.Sub(upload.UploadedAt) < *policyMatch.PolicyDuration {
-					// policyMatch.PolicyID should be non nil here, as any visible uploads cannot be at a tip-of-branch commit
```
Comment (efritz) [inline enterprise/cmd/frontend/internal/codeintel/resolvers/upload_retention.go:132]: Why was comment wrong?

File: enterprise/cmd/frontend/internal/codeintel/resolvers/upload_retention.go [hunk 4]
```diff
@@ -151,12 +154,12 @@ func (r *resolver) populateMatchingCommits(
 	return potentialMatches, potentialMatchIndexSet
 }
 
-func policyByID(policies []dbstore.ConfigurationPolicy, id *int) *dbstore.ConfigurationPolicy {
-	if id == nil {
+func policyByID(policies []dbstore.ConfigurationPolicy, id int) *dbstore.ConfigurationPolicy {
+	if id == -1 {
```
Comment (efritz) [inline enterprise/cmd/frontend/internal/codeintel/resolvers/upload_retention.go:158]: Could take an `id *int` here instead and make it a bit shorter at the callsites

File: enterprise/cmd/frontend/internal/codeintel/resolvers/upload_retention.go (hunk 3 above)
Comment (efritz) [reply enterprise/cmd/frontend/internal/codeintel/resolvers/upload_retention.go:132]: Hey I read the PR description. I'm out. 
image

File: client/web/src/enterprise/codeintel/uploads/components/UploadRetentionStatusNode.tsx (hunk 2 above)
Comment (Strum355) [reply client/web/src/enterprise/codeintel/uploads/components/UploadRetentionStatusNode.tsx:55]: `9` should probably actually be `7`,  cant remember why I set it to `9` :thinking: 
`4` is an arbitrary number so we dont end up listing every single visible commit

**Review by efritz — APPROVED** [review]
Have suggestions but they're non-blocking to get the fix out.

File: client/web/src/enterprise/codeintel/uploads/components/UploadRetentionStatusNode.tsx (hunk 2 above)
Comment (efritz) [reply client/web/src/enterprise/codeintel/uploads/components/UploadRetentionStatusNode.tsx:55]: What are they actually limiting? Can we give these const names as documentation?

Comment (Strum355) [reply client/web/src/enterprise/codeintel/uploads/components/UploadRetentionStatusNode.tsx:55]: the `9` that should be `7` is just truncating the sha to short version like [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@nsc/retentionpolicyoverview-fix/-/blob/client/web/src/enterprise/codeintel/shared/components/CodeIntelUploadOrIndexCommit.tsx?L22%3A60=&utm_source=VSCode-2.0.9)

File: enterprise/cmd/frontend/internal/codeintel/resolvers/upload_retention.go (hunk 4 above)
Comment (Strum355) [reply enterprise/cmd/frontend/internal/codeintel/resolvers/upload_retention.go:158]: It doesnt actually end up saving us any space, as I need to do a nil check _anyways_ to assign the magic `-1` index value for the tip-of-default-branch implicit policy being inserted into `potentialMatchIndexSet` [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@nsc/retentionpolicyoverview-fix/-/blob/enterprise/cmd/frontend/internal/codeintel/resolvers/upload_retention.go?L120%3A5-120%3A27=) and [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@nsc/retentionpolicyoverview-fix/-/blob/enterprise/cmd/frontend/internal/codeintel/resolvers/upload_retention.go?L147&utm_source=VSCode-2.0.9)

**Review by efritz — APPROVED** [review]

**Pull request merged**